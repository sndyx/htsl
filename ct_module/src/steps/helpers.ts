import { Step } from './step';
import { getSlotFromName } from '../slots';
import { normalized, assertIsNormalized } from '../utils';

export function booleanAsValue(value: boolean): string {
    return value ? 'Enabled' : 'Disabled';
}

export function numberAsValue(value: number): string {
    return value.toString();
}

export function stringAsValue(value: string): string {
    return value;
}

export function stepsClickButtonThenSelectValue(
    key: string,
    keyIsNormalized: boolean,
    value: string,
    valueIsNormalized: boolean
): Step[] {
    if (keyIsNormalized) {
        assertIsNormalized(key);
    }
    if (valueIsNormalized) {
        assertIsNormalized(value);
    }
    return [
        {
            type: 'CLICK_BUTTON',
            key: key,
            keyIsNormalized: keyIsNormalized,
        },
        {
            type: 'SELECT_VALUE',
            key: key,
            keyIsNormalized: keyIsNormalized,
            value: value,
            valueIsNormalized: valueIsNormalized,
        },
    ];
}

export function stepsAddAction(actionName: string): Step[] {
    assertIsNormalized(actionName);
    return stepsClickButtonThenSelectValue('add_action', true, actionName, true);
}

export function stepGoBack(): Step {
    return {
        type: 'CLICK_BUTTON',
        key: 'go_back',
        keyIsNormalized: true,
    };
}

export function stepClickButtonOrNextPage(name: string, isNormalized: boolean): Step {
    if (isNormalized) {
        assertIsNormalized(name);
    }
    return {
        type: 'CONDITIONAL',
        condition: () => getSlotFromName(name, isNormalized) !== null,
        then: () => [
            {
                type: 'CLICK_BUTTON',
                key: name,
                keyIsNormalized: isNormalized,
            },
        ],
        else: () => [
            {
                type: 'CONDITIONAL',
                condition: () => getSlotFromName('next_page', true) !== null,
                then: () => [
                    {
                        type: 'CLICK_BUTTON',
                        key: 'next_page',
                        keyIsNormalized: true,
                    },
                    stepClickButtonOrNextPage(name, isNormalized),
                ],
                else: () => {
                    throw new Error(`Could not find slot for key: ${name}`);
                },
            },
        ],
    };
}
