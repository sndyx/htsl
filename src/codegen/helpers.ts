import type { WrittenStyle } from './style';

export function error(message: string): string {
    return `/* error: ${message} */`
}

export function hint(message: string): string {
    return `/* ${message} */`
}

export function concat(elements: string[], sep: string): string {
    let res = elements[0] ?? "";

    for (let i = 1; i < elements.length; i++) {
        res = res.concat(sep, elements[i]);
    }

    return res;
}

export function maybeQuote(name: string): string {
    if (name.includes(' ')) return `"${name}"`;
    return name;
}

export function withWrittenStyle(
    name: string,
    style: WrittenStyle,
): string {
    let capitalizedName;
    if (style.capitalization === 'lowercase') {
        capitalizedName = name.toLowerCase();
    } else if (style.capitalization === 'uppercase') {
        capitalizedName = name.toUpperCase();
    } else {
        capitalizedName = name.toLowerCase().split(' ').map(word => {
            return word.charAt(0).toUpperCase() + word.slice(1);
        }).join(' ');
    }

    if (style.quoted || name.includes(' ')) {
        return `"${capitalizedName}"`;
    } else return capitalizedName;
}