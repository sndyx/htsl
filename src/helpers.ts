import type { Action, Comparison, Condition, Operation } from 'housing-common/src/types';

export const ACTION_KWS = [
    'applyLayout',
    'applyPotion',
    'balanceTeam',
    'cancelEvent',
    'changeHealth',
    'hungerLevel',
    'maxHealth',
    'changePlayerGroup',
    'clearEffects',
    'closeMenu',
    'actionBar',
    'displayMenu',
    'title',
    'enchant',
    'exit',
    'failParkour',
    'fullHeal',
    'xpLevel',
    'giveItem',
    'houseSpawn',
    'kill',
    'parkCheck',
    'pause',
    'sound',
    'removeItem',
    'resetInventory',
    'chat',
    'lobby',
    'compassTarget',
    'gamemode',
    'setTeam',
    'tp',
    'consumeItem',
    'stat',
    'globalstat',
    'teamstat',
    'launch',
    'changeVelocity',
    'dropItem',
    'function',
    'random',
    'if',
    'var',
    'globalvar',
    'teamvar',
] as const;

export type ActionKey = Action["type"];
export type ActionKw = (typeof ACTION_KWS)[number];

export const CONDITION_KWS = [
    'blockType',
    'damageAmount',
    'damageCause',
    'doingParkour',
    'fishingEnv',
    'hasItem',
    'hasPotion',
    'isItem',
    'isSneaking',
    'maxHealth',
    'isFlying',
    'health',
    'hunger',
    'portal',
    'canPvp',
    'gamemode',
    'hasGroup',
    'hasPermission',
    'hasTeam',
    'inRegion',
    'stat',
    'globalstat',
    'teamstat',
    'placeholder',
    'var',
    'globalvar',
    'teamvar',
] as const;

export type ConditionKey = Condition["type"];
export type ConditionKw = (typeof CONDITION_KWS)[number];

export const ACTIONS_TO_KWS: {
    [key in Action['type']]: ActionKw;
} = {
    ACTION_BAR: 'actionBar',
    APPLY_INVENTORY_LAYOUT: 'applyLayout',
    APPLY_POTION_EFFECT: 'applyPotion',
    CANCEL_EVENT: 'cancelEvent',
    CHANGE_HEALTH: 'changeHealth',
    CHANGE_HUNGER: 'hungerLevel',
    CHANGE_MAX_HEALTH: 'maxHealth',
    CHANGE_VAR: 'var',
    CLEAR_POTION_EFFECTS: 'clearEffects',
    CONDITIONAL: 'if',
    DROP_ITEM: 'dropItem',
    ENCHANT_HELD_ITEM: 'enchant',
    EXIT: 'exit',
    FAIL_PARKOUR: 'failParkour',
    FUNCTION: 'function',
    GIVE_EXPERIENCE_LEVELS: 'xpLevel',
    GIVE_ITEM: 'giveItem',
    HEAL: 'fullHeal',
    KILL: 'kill',
    LAUNCH: 'launch',
    MESSAGE: 'chat',
    PAUSE: 'pause',
    PLAY_SOUND: 'sound',
    RANDOM: 'random',
    REMOVE_ITEM: 'removeItem',
    RESET_INVENTORY: 'resetInventory',
    SEND_TO_LOBBY: 'lobby',
    SET_COMPASS_TARGET: 'compassTarget',
    SET_GAMEMODE: 'gamemode',
    SET_GROUP: 'changePlayerGroup',
    SET_MENU: 'displayMenu',
    SET_TEAM: 'setTeam',
    SET_VELOCITY: 'changeVelocity',
    TELEPORT: 'tp',
    TITLE: 'title',
};

export const CONDITIONS_TO_KWS: {
    [key in Condition['type']]: ConditionKw;
} = {
    COMPARE_DAMAGE: 'damageAmount',
    COMPARE_HEALTH: 'health',
    COMPARE_HUNGER: 'hunger',
    COMPARE_MAX_HEALTH: 'maxHealth',
    COMPARE_PLACEHOLDER: 'placeholder',
    COMPARE_VAR: 'var',
    IS_DOING_PARKOUR: 'doingParkour',
    IS_FLYING: 'isFlying',
    IS_IN_REGION: 'inRegion',
    IS_SNEAKING: 'isSneaking',
    REQUIRE_GAMEMODE: 'gamemode',
    REQUIRE_GROUP: 'hasGroup',
    REQUIRE_ITEM: 'hasItem',
    REQUIRE_PERMISSION: 'hasPermission',
    REQUIRE_POTION_EFFECT: 'hasPotion',
    REQUIRE_TEAM: 'hasTeam',
};

export const OPERATION_SYMBOLS: {
    [key in Operation]: string;
} = {
    increment: '+=',
    decrement: '-=',
    set: '=',
    multiply: '*=',
    divide: '/=',
};

export const COMPARISON_SYMBOLS: {
    [key in Comparison]: string;
} = {
    less_than: '<',
    less_than_or_equals: '<=',
    equals: '==',
    greater_than: '>',
    greater_than_or_equals: '>=',
};

export const SHORTHANDS = [
    'globalstat',
    'stat',
    'teamstat',
    'randomint',
    'health',
    'maxHealth',
    'hunger',
    'locX',
    'locY',
    'locZ',
    'unix',
] as const;

export function partialEq(src: any, target: any): boolean {
    return Object.keys(target).every((key) => {
        return target[key] === src[key];
    });
}
