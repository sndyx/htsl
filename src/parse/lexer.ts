import { span } from '../span.js';
import { token, type Token } from './token.js';
import { error } from '../diagnostic.js';

export class Lexer {
    src: string;
    pos: number;

    constructor(src: string) {
        this.src = src;
        this.pos = 0;
    }

    advanceToken(): Token {
        // eat whitespace
        while (this.hasNext() && /^\s+$/.test(this.peek()) && this.peek() != '\n') {
            this.next();
        }
        if (!this.hasNext()) return token('eof', span(this.pos, this.pos));

        const lo = this.pos;
        const singleSpan = span(lo, lo + 1);
        const c = this.next();

        if (c === '/' && this.peek() === '/') {
            // eat line comment
            do {
                this.next();
            } while (this.hasNext() && this.peek() !== '\n');

            return this.advanceToken();
        }

        if (c === '/' && this.peek() === '*') {
            this.next();

            // eat block comment
            let depth = 1;
            while (this.hasNext()) {
                const c = this.next();
                if (c === '/' && this.peek() === '*') {
                    this.next();
                    depth++;
                } else if (c === '*' && this.peek() === '/') {
                    this.next();
                    depth--;
                    if (depth === 0) break;
                }
            }

            return this.advanceToken();
        }

        if (c === ',') return token('comma', singleSpan);
        if (c === '!') return token('exclamation', singleSpan);

        // binary operators
        if (c === '+') {
            if (this.peek(0) === '=') {
                this.next();
                return token('bin_op_eq', span(lo, lo + 2), { op: 'plus' });
            }
            return token('bin_op', singleSpan, { op: 'plus' });
        }
        if (c === '-') {
            if (this.peek(0) === '=') {
                this.next();
                return token('bin_op_eq', span(lo, lo + 2), { op: 'minus' });
            }
            return token('bin_op', singleSpan, { op: 'minus' });
        }
        if (c === '*') {
            if (this.peek(0) === '*') {
                this.next();
                return token('bin_op', span(lo, lo + 2), { op: 'star_star' });
            }
            if (this.peek(0) === '=') {
                this.next();
                return token('bin_op_eq', span(lo, lo + 2), { op: 'star' });
            }
            return token('bin_op', singleSpan, { op: 'star' });
        }
        if (c === '/') {
            if (this.peek(0) === '/') this.next();
            if (this.peek(0) === '=') {
                this.next();
                return token('bin_op_eq', span(lo, this.pos), { op: 'slash' });
            }
            return token('bin_op', span(lo, this.pos), { op: 'slash' });
        }

        // comparison operators
        if (c === '=') {
            if (this.peek(0) === '=') {
                this.next();
                return token('cmp_op_eq', span(lo, lo + 2), { op: 'equals' });
            }
            return token('cmp_op', singleSpan, { op: 'equals' });
        }
        if (c === '<') {
            if (this.peek(0) === '=') {
                this.next();
                return token('cmp_op_eq', span(lo, lo + 2), { op: 'less_than' });
            }
            return token('cmp_op', singleSpan, { op: 'less_than' });
        }
        if (c === '>') {
            if (this.peek(0) === '=') {
                this.next();
                return token('cmp_op_eq', span(lo, lo + 2), { op: 'greater_than' });
            }
            return token('cmp_op', singleSpan, { op: 'greater_than' });
        }

        // delimiters
        if (c === '(') return token('open_delim', singleSpan, { delim: 'parenthesis' });
        if (c === ')') return token('close_delim', singleSpan, { delim: 'parenthesis' });
        if (c === '{') return token('open_delim', singleSpan, { delim: 'brace' });
        if (c === '}') return token('close_delim', singleSpan, { delim: 'brace' });
        if (c === '[') return token('open_delim', singleSpan, { delim: 'bracket' });
        if (c === ']') return token('close_delim', singleSpan, { delim: 'bracket' });

        // literals
        if (c === '"') {
            let value = '';
            let escapeNext = false;
            while (this.hasNext()) {
                const c = this.next();
                if (!escapeNext && c === '"') break;
                if (!escapeNext && c === '\\') {
                    escapeNext = true;
                    continue;
                }
                escapeNext = false;
                value += c;
            }

            return token('str', span(lo, this.pos), { value });
        }

        if (c === '%') {
            let value = '';
            while (this.hasNext()) {
                const c = this.next();
                if (c === '%') break;
                value += c;
            }

            return token('placeholder', span(lo, this.pos), { value });
        }

        if (/[0-9]/.test(c)) {
            let value = c;
            while (this.hasNext()) {
                if (!/[0-9]/.test(this.peek())) break;
                value += this.next();
            }
            if (this.peek() === '.') {
                value += '.';
                this.next();
                while (this.hasNext()) {
                    if (!/[0-9]/.test(this.peek())) break;
                    value += this.next();
                }
                return token('f64', span(lo, this.pos), { value });
            }
            return token('i64', span(lo, this.pos), { value });
        }

        if (/[a-zA-Z_]/.test(c)) {
            let value = c;
            while (this.hasNext()) {
                if (!/[a-zA-Z_/\-0-9.-]/.test(this.peek())) break;
                value += this.next();
            }
            return token('ident', span(lo, this.pos), { value });
        }

        if (c === '\n') return token('eol', singleSpan);

        return token('unknown', singleSpan, { value: c });
    }

    hasNext(): boolean {
        return this.pos < this.src.length;
    }

    next(): string {
        return this.src.charAt(this.pos++);
    }

    peek(skip?: number): string {
        return this.src.charAt(this.pos + (skip ?? 0));
    }
}
