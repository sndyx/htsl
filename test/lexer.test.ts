import { describe, it, expect } from 'vitest';
import { Lexer } from '../src/parse/lexer.js';
import { readCases } from './helpers.js';

describe('Lexer', () => {
    for (const test of readCases(__dirname + '/cases/tokens/')) {
        it(test.name, () => {
            const lexer = new Lexer(test.source);
            const result = [];
            while (lexer.hasNext()) result.push(lexer.advanceToken());

            expect(result).toMatchSnapshot();
        });
    }
});
