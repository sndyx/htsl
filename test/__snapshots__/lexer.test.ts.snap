// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lexer > comments 1`] = `
[
  {
    "kind": "ident",
    "span": Span {
      "end": 4,
      "start": 0,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 6,
      "start": 5,
    },
    "value": "x",
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 8,
      "start": 7,
    },
  },
  {
    "kind": "i64",
    "span": Span {
      "end": 10,
      "start": 9,
    },
    "value": "5",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 11,
      "start": 10,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 15,
      "start": 11,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 17,
      "start": 16,
    },
    "value": "y",
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 19,
      "start": 18,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 24,
      "start": 20,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 26,
      "start": 25,
    },
    "value": "x",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 41,
      "start": 40,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 42,
      "start": 41,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 61,
      "start": 60,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 65,
      "start": 61,
    },
    "value": "chat",
  },
  {
    "kind": "str",
    "span": Span {
      "end": 83,
      "start": 66,
    },
    "value": "%stat.player/y%",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 84,
      "start": 83,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 85,
      "start": 84,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 128,
      "start": 127,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 129,
      "start": 128,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 133,
      "start": 129,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 135,
      "start": 134,
    },
    "value": "z",
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 137,
      "start": 136,
    },
  },
  {
    "kind": "i64",
    "span": Span {
      "end": 140,
      "start": 138,
    },
    "value": "10",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 141,
      "start": 140,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 142,
      "start": 141,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 202,
      "start": 201,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 203,
      "start": 202,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 207,
      "start": 203,
    },
    "value": "chat",
  },
  {
    "kind": "str",
    "span": Span {
      "end": 225,
      "start": 208,
    },
    "value": "%stat.player/z%",
  },
]
`;

exports[`Lexer > identifiers 1`] = `
[
  {
    "kind": "ident",
    "span": Span {
      "end": 4,
      "start": 0,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 17,
      "start": 5,
    },
    "value": "myIdentifier",
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 19,
      "start": 18,
    },
  },
  {
    "kind": "i64",
    "span": Span {
      "end": 21,
      "start": 20,
    },
    "value": "5",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 22,
      "start": 21,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 26,
      "start": 22,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 51,
      "start": 27,
    },
    "value": "my/structured/identifier",
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 53,
      "start": 52,
    },
  },
  {
    "kind": "i64",
    "span": Span {
      "end": 56,
      "start": 54,
    },
    "value": "10",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 57,
      "start": 56,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 58,
      "start": 57,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 62,
      "start": 58,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 75,
      "start": 63,
    },
    "value": "myIdentifier",
  },
  {
    "kind": "bin_op_eq",
    "op": "slash",
    "span": Span {
      "end": 78,
      "start": 76,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 83,
      "start": 79,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 108,
      "start": 84,
    },
    "value": "my/structured/identifier",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 109,
      "start": 108,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 113,
      "start": 109,
    },
    "value": "stat",
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 134,
      "start": 114,
    },
    "value": "now_with_underscores",
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 136,
      "start": 135,
    },
  },
  {
    "kind": "i64",
    "span": Span {
      "end": 140,
      "start": 137,
    },
    "value": "100",
  },
]
`;

exports[`Lexer > tokens 1`] = `
[
  {
    "kind": "comma",
    "span": Span {
      "end": 1,
      "start": 0,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 2,
      "start": 1,
    },
  },
  {
    "kind": "bin_op",
    "op": "plus",
    "span": Span {
      "end": 3,
      "start": 2,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 4,
      "start": 3,
    },
  },
  {
    "kind": "bin_op",
    "op": "minus",
    "span": Span {
      "end": 5,
      "start": 4,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 6,
      "start": 5,
    },
  },
  {
    "kind": "bin_op",
    "op": "star",
    "span": Span {
      "end": 7,
      "start": 6,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 8,
      "start": 7,
    },
  },
  {
    "kind": "bin_op",
    "op": "slash",
    "span": Span {
      "end": 9,
      "start": 8,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 10,
      "start": 9,
    },
  },
  {
    "kind": "bin_op",
    "op": "star_star",
    "span": Span {
      "end": 12,
      "start": 10,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 13,
      "start": 12,
    },
  },
  {
    "kind": "bin_op_eq",
    "op": "plus",
    "span": Span {
      "end": 15,
      "start": 13,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 16,
      "start": 15,
    },
  },
  {
    "kind": "bin_op_eq",
    "op": "minus",
    "span": Span {
      "end": 18,
      "start": 16,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 19,
      "start": 18,
    },
  },
  {
    "kind": "bin_op_eq",
    "op": "star",
    "span": Span {
      "end": 21,
      "start": 19,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 22,
      "start": 21,
    },
  },
  {
    "kind": "bin_op_eq",
    "op": "slash",
    "span": Span {
      "end": 24,
      "start": 22,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 25,
      "start": 24,
    },
  },
  {
    "kind": "cmp_op",
    "op": "greater_than",
    "span": Span {
      "end": 26,
      "start": 25,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 27,
      "start": 26,
    },
  },
  {
    "kind": "cmp_op",
    "op": "less_than",
    "span": Span {
      "end": 28,
      "start": 27,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 29,
      "start": 28,
    },
  },
  {
    "kind": "cmp_op",
    "op": "equals",
    "span": Span {
      "end": 30,
      "start": 29,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 31,
      "start": 30,
    },
  },
  {
    "kind": "cmp_op_eq",
    "op": "greater_than",
    "span": Span {
      "end": 33,
      "start": 31,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 34,
      "start": 33,
    },
  },
  {
    "kind": "cmp_op_eq",
    "op": "less_than",
    "span": Span {
      "end": 36,
      "start": 34,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 37,
      "start": 36,
    },
  },
  {
    "kind": "cmp_op_eq",
    "op": "equals",
    "span": Span {
      "end": 39,
      "start": 37,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 40,
      "start": 39,
    },
  },
  {
    "delim": "parenthesis",
    "kind": "open_delim",
    "span": Span {
      "end": 41,
      "start": 40,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 42,
      "start": 41,
    },
  },
  {
    "delim": "brace",
    "kind": "open_delim",
    "span": Span {
      "end": 43,
      "start": 42,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 44,
      "start": 43,
    },
  },
  {
    "delim": "bracket",
    "kind": "open_delim",
    "span": Span {
      "end": 45,
      "start": 44,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 46,
      "start": 45,
    },
  },
  {
    "delim": "parenthesis",
    "kind": "close_delim",
    "span": Span {
      "end": 47,
      "start": 46,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 48,
      "start": 47,
    },
  },
  {
    "delim": "brace",
    "kind": "close_delim",
    "span": Span {
      "end": 49,
      "start": 48,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 50,
      "start": 49,
    },
  },
  {
    "delim": "bracket",
    "kind": "close_delim",
    "span": Span {
      "end": 51,
      "start": 50,
    },
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 52,
      "start": 51,
    },
  },
  {
    "kind": "str",
    "span": Span {
      "end": 60,
      "start": 52,
    },
    "value": "string",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 61,
      "start": 60,
    },
  },
  {
    "kind": "placeholder",
    "span": Span {
      "end": 74,
      "start": 61,
    },
    "value": "placeholder",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 75,
      "start": 74,
    },
  },
  {
    "kind": "i64",
    "span": Span {
      "end": 80,
      "start": 75,
    },
    "value": "12345",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 81,
      "start": 80,
    },
  },
  {
    "kind": "f64",
    "span": Span {
      "end": 87,
      "start": 81,
    },
    "value": "1.2345",
  },
  {
    "kind": "eol",
    "span": Span {
      "end": 88,
      "start": 87,
    },
  },
  {
    "kind": "ident",
    "span": Span {
      "end": 98,
      "start": 88,
    },
    "value": "identifier",
  },
]
`;
